mist-core/src/main/java/edu/snu/mist/core/master/recovery/SingleNodeRecoveryScheduler.java

```diff
@@ -0,0 +1,141 @@
+import edu.snu.mist.formats.avro.GroupStats;
+import edu.snu.mist.formats.avro.MasterToTaskMessage;
+import org.apache.avro.AvroRemoteException;
+
+import javax.inject.Inject;
+import java.util.*;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.locks.Condition;
+import java.util.concurrent.locks.Lock;
+import java.util.concurrent.locks.ReentrantLock;
+import java.util.logging.Level;
+import java.util.logging.Logger;
+
+/**
+ * The recovery manager which leverages only a single node in recovery process.
+ */
+public final class SingleNodeRecoveryScheduler implements RecoveryScheduler {
+
+  private static final Logger LOG = Logger.getLogger(RecoveryScheduler.class.getName());
+
+  /**
+   * The map which contains all the failed groups.
+   */
+  private ConcurrentMap<String, GroupStats> allFailedGroups;
+
+  /**
+   * The map which contains the groups to be recovered.
+   */
+  private ConcurrentMap<String, GroupStats> recoveryGroups;
+
+  /**
+   * The shared taskStatsMap.
+   */
+  private final TaskStatsMap taskStatsMap;
+
+  /**
+   * The shared map which contains Avro proxies to task.
+   */
+  private final ProxyToTaskMap proxyToTaskMap;
+
+  /**
+   * The lock which is used for conditional variable.
+   */
+  private final Lock lock;
+
+  /**
+   * The conditional variable which synchronizes the recovery process.
+   */
+  private final Condition recoveryFinished;
+
+  @Inject
+  private SingleNodeRecoveryScheduler(
+      final TaskStatsMap taskStatsMap,
+      final ProxyToTaskMap proxyToTaskMap) {
+    this.allFailedGroups = new ConcurrentHashMap<>();
+    this.recoveryGroups = null;
+    this.taskStatsMap = taskStatsMap;
+    this.proxyToTaskMap = proxyToTaskMap;
+    this.lock = new ReentrantLock();
+    this.recoveryFinished = this.lock.newCondition();
+  }
+
+  @Override
+  public void recover(final Map<String, GroupStats> failedGroups) {
+    LOG.log(Level.INFO, "Start recovery on failed groups: {0}", failedGroups.keySet());
+    this.allFailedGroups.putAll(failedGroups);
+    MasterToTaskMessage proxyToRecoveryTask;
+    try {
+      lock.lock();
+      this.recoveryGroups = allFailedGroups;
+      this.allFailedGroups = new ConcurrentHashMap<>();
+      // Select the newly allocated task with the least load for recovery
+      double minLoad = Double.MAX_VALUE;
+      proxyToRecoveryTask = null;
+      for (final Map.Entry<String, MasterToTaskMessage> entry : proxyToTaskMap.entrySet()) {
+        final String taskHostname = entry.getKey();
+        final double taskLoad = taskStatsMap.get(taskHostname).getTaskLoad();
+        if (taskLoad < minLoad) {
+          minLoad = taskLoad;
+          proxyToRecoveryTask = entry.getValue();
+        }
+      }
+      if (proxyToRecoveryTask == null) {
+        throw new IllegalStateException("Internal error : ProxyToRecoveryTask shouldn't be null!");
+      }
+      proxyToRecoveryTask.startRecovery();
```

Comment:
Maybe it would be good to change the method name. How about `pullingRecoverableGroups`?